nni/contrib/compression/distillation/basic_distiller.py

```diff
@@ -0,0 +1,277 @@
+
+    @classmethod
+    def from_compressor(cls, compressor: Compressor, new_config_list: List[Dict], teacher_model: torch.nn.Module,
+                        teacher_predict: Callable[[Any, torch.nn.Module], torch.Tensor], origin_loss_lambda: float = 1.,
+                        evaluator: Evaluator | None = None):
+        return super().from_compressor(compressor, new_config_list, evaluator=evaluator, teacher_model=teacher_model,
+                                       teacher_predict=teacher_predict, origin_loss_lambda=origin_loss_lambda)
+
+    def _set_default_link(self):
+        for module_name, ts in self._target_spaces.items():
+            for _, target_space in ts.items():
+                link = target_space.link if target_space.link is not None else 'auto'
+                link = module_name if link == 'auto' else link
+                link = [link] if isinstance(link, str) else link
+                assert all(l in self._teacher_target_spaces for l in link), '`link` should be a module name in teacher model.'
+                target_space.link = link
+
+    def _set_default_lambda(self):
+        for _, ts in self._target_spaces.items():
+            for _, target_space in ts.items():
+                target_space.lambda_ = target_space.lambda_ if target_space.lambda_ is not None else 1.
+
+    def wrap_teacher_model(self):
+        """
+        Traverse all teacher wrappers and execute ModuleWrapper.wrap()
+        """
+        if self._teacher_is_wrapped is True:
+            warn_msg = 'The bound model has been wrapped, no need to wrap again.'
+            _logger.warning(warn_msg)
+        for _, wrapper in self._teacher_module_wrappers.items():
+            wrapper.wrap()
+        self._teacher_is_wrapped = True
+
+    def unwrap_teacher_model(self):
+        """
+        Traverse all teacher wrappers and execute ModuleWrapper.unwrap()
+        """
+        if self._teacher_is_wrapped is False:
+            warn_msg = 'The bound model is not wrapped, can not unwrap it.'
+            _logger.warning(warn_msg)
+        for _, wrapper in self._teacher_module_wrappers.items():
+            wrapper.unwrap()
+        self._teacher_is_wrapped = False
+
+    def _register_loss_patch(self, evaluator: Evaluator):
+        def loss_patch(original_loss, batch):
+            with torch.no_grad():
+                self.teacher_predict(batch, self.teacher_model)
+            return self.origin_loss_lambda * original_loss + self.compute_distill_loss()
+
+        evaluator.patch_loss(loss_patch)
+
+    def compute_distill_loss(self):
+        raise NotImplementedError()
+
+    def _single_compress(self, max_steps: int | None, max_epochs: int | None):
+        self._fusion_compress(max_steps, max_epochs)
+
+    def _fuse_preprocess(self, evaluator: Evaluator):
+        self._register_loss_patch(evaluator)
+
+    def _fuse_postprocess(self, evaluator: Evaluator):
+        pass
+
+
+class DynamicLayerwiseDistiller(TeacherModelBasedDistiller):
```

Comment:
What is `DynamicLayerwiseDistiller`?

Reply:
Each student layer corresponds to a set of teacher layers, during distillation, the student layer will compute a set of distil losses with each teacher layer in its teacher set, then choose the smallest loss in the loss set as this student layer distil loss.

I will add comments for this distiller.